docs/design/constraints.md

```diff
@@ -0,0 +1,177 @@
+New property `Constraints` will be added to `ITemplateInfo` of type `IReadOnlyList<TemplateConstraintInfo>` containing template constraints information.
+
+`TemplateConstraintInfo` has the following properties:
+- `string Type`: constraint type
+- `string Args`: JSON string containing args read from `template.json`. Args to be parsed when evaluating the constraint by constraint itself.
+
+### Types uniqueness consideration
+At the moment 3rd party constraints won't be supported. 1st party developers should ensure unique naming.
+When 3rd party constraints are supported, the way to check if the constraint type is already used should be considered.
+
+
+## Constraint component
+
+Constraint implementation should implement `ITemplateConstraintFactory` interface that creates `ITemplateConstraint`:
+```csharp
+    public interface ITemplateConstraintFactory : IIdentifiedComponent
+    {
+        /// <summary>
+        /// Unique constraint type. Should match the definition in `template.json`.
+        /// </summary>
+        string Type { get; }
+
+        /// <summary>
+        /// Creates new <see cref="ITemplateConstraint"/> based on current <see cref="IEngineEnvironmentSettings"/>.
+        /// </summary>
+        public Task<ITemplateConstraint> CreateTemplateConstraintAsync(IEngineEnvironmentSettings environmentSettings);
+    }
+```
+When creating the constraint, it is assumed that constraint will be initialized and its context will be evaluated.
+Example: if constraint needs to know the project MSBuild properties it is assumed that project will be identified and MSBuild properties will be evaluated when the constraint is being created prior to first evaluation.
+Template constraints initialization will be started in parallel once `IEngineEnvironmentSettings` are created. The methods that work with constraint will ensure task completion before evaluation.
+
+`ITemplateConstraint` definition:
+```csharp
+    public interface ITemplateConstraint
+    {
+        /// <summary>
+        /// Unique constraint type. Should match the definition in `template.json`.
+        /// </summary>
+        string Type { get; }
+
+        /// <summary>
+        /// The user friendly constraint name, that can be used in UI.
+        /// </summary>
+        string DisplayName { get; }
+
+        /// <summary>
+        /// Evaluates <see cref="ITemplateInfo"/> and returns <see cref="TemplateConstraintResult"/> containing result.
+        /// </summary>
+        TemplateConstraintResult Evaluate(string args);
+    }
+
+```
+where `args` contains arguments for constraint read from `template.json`. `Type` matches `type` from constraint configuration (case-sensitive).
+Note: the parameter type may be extended if needed. 
+`Evaluate` method should have optimal performance and operate on the context already evaluated when the constraint was created. It should not launch any time-consuming operation.
```

Comment:
Like mentioned, above, that may not always be possible, since args are not available at the time of constraint creation.